kafka/consumer.go

```diff
@@ -249,3 +242,143 @@ func (c *Consumer) Healthy() error {
 	}
 	return nil
 }
+
+// consumer wraps partitionConsumers and exposes the necessary callbacks
+// to use when partitions are reassigned.
+type consumer struct {
+	mu        sync.Mutex
+	wg        sync.WaitGroup
+	consumers map[topicPartition]partitionConsumer
+	processor model.BatchProcessor
+	logger    *zap.Logger
+	decoder   Decoder
+	delivery  apmqueue.DeliveryType
+}
+
+type topicPartition struct {
+	topic     string
+	partition int32
+}
+
+// assigned must be set as a kgo.OnPartitionsAssigned callback. Ensuring all
+// assigned partitions to this consumer process received records.
+func (c *consumer) assigned(_ context.Context, client *kgo.Client, assigned map[string][]int32) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for topic, partitions := range assigned {
+		for _, partition := range partitions {
+			c.wg.Add(1)
+			pc := partitionConsumer{
+				wg:        &c.wg,
+				records:   make(chan []*kgo.Record),
+				processor: c.processor,
+				logger:    c.logger,
+				decoder:   c.decoder,
+				client:    client,
+				delivery:  c.delivery,
+			}
+			go pc.consume(topic, partition)
```

Comment:
```suggestion
			go func(topic string, partition int32) {
				defer c.wg.Done()
				pc.consume(topic, partition)
			}(topic, partition)
```

Then the partition consumer doesn't need to know about the wait group, and it's kept close to where you call Add.